### PR TITLE
Reimplement some builder integration tests as unit tests

### PR DESCRIPTION
**Pull request by tkopczynski**
Moving some builder tests from integration-cli to unit tests.

Part of resolving #20784

Signed-off-by: Tomasz Kopczynski tomek@kopczynski.net.pl

### CODE REVIEW

**Comment by vdemeester** [comment]
LGTM :+1:

File: builder/dockerfile/evaluator_test.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+	tarStream, err := archive.Tar(contextDir, archive.Uncompressed)
+
+	if err != nil {
+		t.Fatalf("Error when creating tar stream: %s", err)
+	}
+
+	defer tarStream.Close()
+
+	context, err := builder.MakeTarSumContext(tarStream)
```
Comment (runcom) [inline builder/dockerfile/evaluator_test.go]: check err

File: builder/dockerfile/evaluator_test.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+	r := strings.NewReader(testCase.dockerfile)
+	n, err := parser.Parse(r)
```
Comment (runcom) [inline builder/dockerfile/evaluator_test.go]: check err

File: builder/dockerfile/evaluator_test.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	config := &container.Config{}
+	options := &types.ImageBuildOptions{}
+
+	b := &Builder{runConfig: config, options: options, Stdout: ioutil.Discard, context: context}
+
+	err = b.dispatch(0, n.Children[0])
+
+	if err == nil {
+		t.Fatalf("No error when executing test %s", testCase.name)
+	} else {
```
Comment (runcom) [inline builder/dockerfile/evaluator_test.go]: no need for this `else`

**Comment by runcom** [comment]
small nits, LGTM otherwise

**Comment by tkopczynski** [comment]
hey @runcom, I just corrected errors that you mentioned, ptal

**Comment by runcom** [comment]
looks fine to me now

File: builder/dockerfile/evaluator_test.go (hunk 1 above)
Comment (runcom) [inline builder/dockerfile/evaluator_test.go]: should you `defer context.Close`? (not sure)

**Comment by tkopczynski** [comment]
@runcom you were right, I found closing this context in the builder code. I updated the PR.

**Pull request merged**